testbed/tests/window/test_window.py

```diff
@@ -507,6 +662,133 @@ async def test_move_and_resize(second_window, second_window_probe):
         assert second_window.size == (250 + extra_width, 210 + extra_height)
         assert second_window_probe.content_size == (250, 210)
 
+    @pytest.mark.parametrize(
+        "initial_state, final_state",
+        [
+            # Switch from NORMAL:
+            (WindowState.NORMAL, WindowState.MINIMIZED),
+            (WindowState.NORMAL, WindowState.MAXIMIZED),
+            (WindowState.NORMAL, WindowState.FULLSCREEN),
+            (WindowState.NORMAL, WindowState.PRESENTATION),
+            (WindowState.NORMAL, WindowState.NORMAL),
+            # Switch from MINIMIZED:
+            (WindowState.MINIMIZED, WindowState.NORMAL),
+            (WindowState.MINIMIZED, WindowState.MAXIMIZED),
+            (WindowState.MINIMIZED, WindowState.FULLSCREEN),
+            (WindowState.MINIMIZED, WindowState.PRESENTATION),
+            (WindowState.MINIMIZED, WindowState.MINIMIZED),
+            # Switch from MAXIMIZED:
+            (WindowState.MAXIMIZED, WindowState.NORMAL),
+            (WindowState.MAXIMIZED, WindowState.MINIMIZED),
+            (WindowState.MAXIMIZED, WindowState.FULLSCREEN),
+            (WindowState.MAXIMIZED, WindowState.PRESENTATION),
+            (WindowState.MAXIMIZED, WindowState.MAXIMIZED),
+            # Switch from FULLSCREEN:
+            (WindowState.FULLSCREEN, WindowState.NORMAL),
+            (WindowState.FULLSCREEN, WindowState.MINIMIZED),
+            (WindowState.FULLSCREEN, WindowState.MAXIMIZED),
+            (WindowState.FULLSCREEN, WindowState.PRESENTATION),
+            (WindowState.FULLSCREEN, WindowState.FULLSCREEN),
+            # Switch from PRESENTATION:
+            (WindowState.PRESENTATION, WindowState.NORMAL),
+            (WindowState.PRESENTATION, WindowState.MINIMIZED),
+            (WindowState.PRESENTATION, WindowState.MAXIMIZED),
+            (WindowState.PRESENTATION, WindowState.FULLSCREEN),
+            (WindowState.PRESENTATION, WindowState.PRESENTATION),
+        ],
+    )
+    @pytest.mark.parametrize(
+        "intermediate_states",
+        [
+            # These sets of intermediate states are specifically chosen to trigger cases that
+            # will cause test failures if the implementation is incorrect on certain backends,
+            # such as macOS.
```

Comment:
We've now removed the mobile versions of these intermediate tests - but it's not clear to me why we need *6* states to test an intermediate transition. An intermediate state transition involves *3* states - the start state, the end state, and an intermediate state that needs to be honoured in some way. As I've mentioned in a couple of comments, the "full" test suite would be 125 "3 state" transitions, but that would be prohibitively complex. Why do we need chains of *6* states here?

Reply:
The need for the chain of 6 states is to trigger the logic for handling of newly requested states when a previous state is pending for application.

Currently, the test logic is like:
* Assign initial window state to window
|
* Wait for window 
|
* Assert window state is initial state
|
* Assign chain of 6 window states
|
* Assign the final window state
| 
* Wait for window
|
* Assert window state is final state

For example:
* Initially =>`pending_state = None`
* FULLSCREEN is requested, but pending for application => pending_state = FULLSCREEN, since the window state was NORMAL 
* MINIMIZED is requested, but FULLSCREEN is not applied => pending_state = MINIMIZED
* PRESENTATION is requested, but FULLSCREEN is not applied => pending_state = PRESENTATION
* Now, FULLSCREEN is applied, and it checks for pending_state i.e., PRESENTATION
* Since, current_state(FULLSCREEN) != pending_state(PRESENTATION), so it applies NORMAL 
* When NORMAL state is reached, it then applies pending_state(PRESENTATION)

Essentially, it triggers the code logic branch: "Request to apply new state when there is a pending_state"

If there were only 1 or 2 intermediate states, then the code branches for "applying new window state when there is a pending state would not be triggered".

For example, 
https://github.com/beeware/toga/blob/7e7cc1e54f24daa8b91c957c92ba68451b14d96f/cocoa/src/toga_cocoa/window.py#L87-L100
In the above, the 6 chain intermediate states trigger the branch:
https://github.com/beeware/toga/blob/7e7cc1e54f24daa8b91c957c92ba68451b14d96f/cocoa/src/toga_cocoa/window.py#L89-L98
But if there were only 1 or 2 intermediate states then it would only trigger the following branch:
https://github.com/beeware/toga/blob/7e7cc1e54f24daa8b91c957c92ba68451b14d96f/cocoa/src/toga_cocoa/window.py#L99-L100

Reply:
I understand that these are the lines of code you are trying to cover, and that sequence of six transitions is *one* mechanism by which you can cover these lines of code. The point I'm trying to make (and this is a recurring theme in reviews of your code) is that the purpose of this exercise is *not* to make the test suite stop complaining about coverage. It's to ensure that the code is completely tested, *and* - and this is the important part - that it's *obvious* why the code is completely tested.

A test should be as *simple* as possible, as *atomic* as possible. This sometimes requires that there are multiple tests. The simplicity and atomicity is an *essential* part of the process, because it's part of making future bugs easy to identify.

As an example - let's say a future developer decides the `performSelector` call on entering fullscreen is clearly redundant, and "simplifies" the code. This test suite will report the failure, but it won't be clear *why* the test has failed. A sequence of 6 state changes, many of which aren't FULLSCREEN, will suddenly stop working. This will then require a complex debugging process to work out why that *specific* test is failing. 

However, if the test is a sequence of 3 that clearly involves entering FULLSCREEN mode, it will be a lot more clear where the problem lies.

> * Initially =>`pending_state = None`
> * FULLSCREEN is requested, but pending for application => pending_state = FULLSCREEN, since the window state was NORMAL
> * MINIMIZED is requested, but FULLSCREEN is not applied => pending_state = MINIMIZED
> * PRESENTATION is requested, but FULLSCREEN is not applied => pending_state = PRESENTATION

My core point - this part of the sequence is already describing *two distinct intermediate tests*:

* requesting MINIMIZED when FULLSCREEN has not been applied
* requesting PRESENTATION when FULLSCREEN has not been applied.
 
As written, the test to this point is a test of "PRESENTATION has been requested when a MINIMIZED state has been requested but not applied because FULLSCREEN hasn't completed". So - if the test fails... is the problem with the MINIMIZED request or the PRESENTATION request? 

If the state logic were so complex that this was, in fact, a problematic transition, then yes - it might need to be tested as a longer sequence - but that would also be a flag that there's a much bigger problem with the underlying logic - or, at least, that there's a need for a specific explanation in the docstring associated with the test to clarify the specific set of circumstances, not just "this is a state change known to be a problem".

Reply:
I agree with you, and looking at the tests, I now realize that I have essentially mashed together 2 different tests into one. Thereby, making it harder to follow the test logic and debug on failure. I will separate these tests and report back.